### ISSUE TITLE
Don't spread the rest arguments unless requested

### ISSUE THREAD

**Issue by Jamesernator** [rbuckton/proposal-partial-application#4]
Now the current proposal lets the additional arguments passed to the function be spread into the calling function but this might cause surprising semantics  for anywhere that receives multiple arguments (particularly array methods) e.g.:

```js
[11,22,33].forEach(console.log(?))
// Prints (in Node.js anyway)
// 11 0 [11, 22, 33]
// 22 1 [11, 22, 33]
// 33 2 [11, 22, 33]
// I probably would've expected just the first argument to be considered


// Resolves with the reject function as resolution value
// not terrible a bit weird to say the least
const delay = new Promise(setTimeout(?, 1000))

// It's also unavoidable for reject to be one of the arguments to a callback
// with setTimeout e.g.
const delayer = setTimeout(?, 1000, 'banana')

delayer((...args) => { console.log(args) })
// ['banana', <reject function>]

// Probably won't work correctly with the majority of existing popular
// libraries e.g. these won't work with lodash

[[1,2], [3,4], [5,6], [7,8]].map(_.concat([99], ?))
// would become [ [ 99, 1, 2, 0, [ 1, 2 ], [ 3, 4 ], [ 5, 6 ], [ 7, 8 ] ],
//  [ 99, 3, 4, 1, [ 1, 2 ], [ 3, 4 ], [ 5, 6 ], [ 7, 8 ] ],
//  [ 99, 5, 6, 2, [ 1, 2 ], [ 3, 4 ], [ 5, 6 ], [ 7, 8 ] ],
//  [ 99, 7, 8, 3, [ 1, 2 ], [ 3, 4 ], [ 5, 6 ], [ 7, 8 ] ] ]
// which is almost definitely not intended

[['my', 'csv', 'data'],
 ['my', 'extra', 'data'],
 ['and', 'more', 'data']
].map(_.slice(?, 2))
// would become [[], [], []] because the "end" to slice to is a string

// etc I think my point is clear from those examples
```

Given that virtually every popular library won't be usable with the syntax in places like `.map`/`.filter`/`new Promise`/etc I'd much much rather see spread as opt in e.g.:

```js
function add(...args) {
    return args.reduce((acc, arg) => acc + arg, 0)
}

[1,2,3,4,5].map(add(2, ?)) // [3,4,5,6,7]
[1,2,3,4,5].map(add(2, ?...)) // Opt in explicitly 
// ["11,2,3,4,5","31,2,3,4,5","51,2,3,4,5","71,2,3,4,5","91,2,3,4,5"]
// if they *actually* wanted that

function repeatByIndex(joiner, string, n) {
    return new Array(n).fill(string).join(joiner)
}

['fizz', 'buzz', 'baz', 'qux'].map(repeatByIndex('@@', ?...))
// ['', 'buzz', 'baz@@baz', 'qux@@qux@@qux']
```

**Comment by rbuckton** [comment]
It's currently commented out in the strawman, but I have considered just using `...` as the placeholder rest/spread token meaning "put the rest of the arguments here."

**Comment by rbuckton** [comment]
Fixed by 8cbd1c0f6c9e6e115211913de8cfd5464a6cf54b.
